### DISCUSSION TITLE
LibreOffice performance with Gotenberg 7.4.x

### DISCUSSION THREAD

**Discussion by gulien** [gotenberg/gotenberg#391] [Q&A]
A discussion to monitor and gather feedbacks from the community about the performances of the LibreOffice module.

Initial discussion may be found in #373.

**Comment by omni-htg** [comment]
First of all, thank you _so_ much for managing to fit in this huge change in-between the family (congrats!).
Quick question would be:
How decent of a solution would it be to stop the container and start it over every X amount of time in order to deal with LibreOffice's memory leaks?

Thanks a bunch.

**Reply by gulien** [reply]
Hey thanks @omni-htg! My daughter is actually quite calm so it is not that hard ; after all the chores at home are done, I do have quite a lot of free time 😅

Regarding the memory leak, it seems non-existent (at least, I do not see it in the live demo). AFAIK, the unoconv listener seems to stop the LibreOffice instance after a while, because the first conversion take more time than those just after. I’m not 100% confident about this statement though.

**Reply by gulien** [reply]
@HadrienKulik @sveisvei are also doing some experiments on their side, maybe they’ll bring their insights about this subject 😄

**Reply by omni-htg** [reply]
That's some _great_ news, both on your daughter and the leaks!
With unoserver being a thing, I wasn't really sure that unoconv had a start-stop behaviour by itself -- so that's pretty nice.
If you haven't run into problems with a 24/7 container, I hope we can get a similar experience on our side.

**Reply by gulien** [reply]
I’m not sure there is a start-stop mechanism in unoconv, but, as you said, the live demo proves that it seems to work just fine 😊. Anyway, it’s always possible to rollback to the previous behavior with `—unoconv-disable-listener` 😄

**Reply by omni-htg** [reply]
Indeed, that was a great addition.
The thing is that, similar to @HadrienKulik, this "listener" style behaviour is great for multiple / batch conversions, so it wouldn't be the best of choices to go back haha

Since I'm here (and if that's OK with you), mind if I ask:
Would you happen to know how I could add the "pdf export option" "PDFUACompliance" to gotenberg / unoconv?
I've tried "sending" it as one of the -e parameters in unoconv, but no dice.

**Reply by gulien** [reply]
No idea, but please open an issue with your use cases so that I can have a better understanding :)

**Comment by wedi** [comment]
Do you have any real world numbers of Gotenbergs docx to pdf performance? I'm not looking for scientifically proven numbers but a rough idea of what's possible.

I am planning to use Gotenberg from a Laravel Worker. At times, there will be batches of a three digit number of PDFs to create.  Throttling could probably be handled from the Laravel side by checking the queue length before requesting the next conversion.

Unfortunately, I have not yet understood how/if queuing works. Somewhere I read that there is no queuing and requests will time out. Somewhere else I read that you can ask for the current queue size. One explanation could be that requests are synchronous but in that case I don't see the benefit of having webhooks. So I guess there is some queuing?  

It would be nice if you could shed some light on this @gulien. Thank you!

**Reply by gulien** [reply]
Hello @wedi,

Actually, there are two ways to setup de LibreOffice module:

* (**default**) A long-running LibreOffice instance, for faster unitary conversions but that can only perform one conversion at a time (with a risk of timeout if the queue is not unstacking fast enough).
* Stateless (`--uno-listener-restart-threshold=0`), i.e., a dedicated LibreOffice instance per conversion (the number of concurrent requests will be limited by the CPU/Memory). 

The queue length is a Prometheus metric that makes sense mostly for monitoring purposes. Also, in stateless mode, the queue length is always 0 😄

**Reply by wedi** [reply]
Thanks for your quick reply!

I'm here to find out if Gotenberg is able to handle batches of 100s of PDFs, and how if it is.

At first sight, it doesn't make a lot of sense to me to start and stop a whole LibreOffice instance for every PDF if you do many at once. So, if you want to get it done in a reasonable time, the stateful version it is. I understand that one instance can only handle one document at a time.

I still have not understood how queuing works in stateful mode. So let's assume, we have a batch of 100 documents to convert.  

My naive approach would be to get a beefy machine, run `docker-compose up --scale gotenberg=8`, fire 10 curl requests for asynchronous operation to each instance and wait for the PDFs to drop in via webhooks. Where exactly does the timeout occur? I had expected a timeout only for synchronous requests...

**Reply by gulien** [reply]
The timeout occurs if a request takes more than a certain amount of time to finish (default is 30 seconds), including asynchronous workflows.

Let’s say you send 10 concurrent requests:

* In stateful configuration, each requests will lock, one at a time, the LibreOffice instance. So if a conversion takes 2sec on average, it will take the whole operation around 20 seconds in our example.
* In stateless configuration, if a conversion takes 5 sec, the whole operation will take a little bit more than 5 seconds (but consumes more CPU/memory).

My guess is that the stateless configuration works better for batch conversions, but your best option is to test both and see what works best for you ☺️
